src/Components/Components/src/Rendering/ComponentState.cs

```diff
@@ -222,5 +222,27 @@ private void DisposeBuffers()
             ((IDisposable)CurrentRenderTree).Dispose();
             _latestDirectParametersSnapshot?.Dispose();
         }
+
+        internal bool RequiresAsyncDisposal() => Component is IAsyncDisposable;
+
+        internal async Task DisposeInBatchAsync(RenderBatchBuilder batchBuilder)
+        {
+            _componentWasDisposed = true;
+
+            // We don't expect these things to throw.
+            RenderTreeDiffBuilder.DisposeFrames(batchBuilder, CurrentRenderTree.GetFrames());
+
+            if (_hasAnyCascadingParameterSubscriptions)
+            {
+                RemoveCascadingParameterSubscriptions();
+            }
+
+            DisposeBuffers();
+
+            if (Component is IAsyncDisposable asyncDisposable)
+            {
+                await asyncDisposable.DisposeAsync();
```

Comment:
There's no other `await` in this method so could directly `return` this task?

Reply:
I want to await here to avoid having to deal separately between sync exceptions and async exceptions in the caller, I don't think this will be common enough that avoiding the allocation will produce a meaningful difference in perf.

Reply:
I ended up avoiding the allocation